Fix `blob.service_url` for supports string type `:filename` option

huacnlee opened the pull request:
Fix `blob.service_url` for supports string or nil `:filename` option, and make sure that present a `ActiveStorage::Filename` type to `serivce.url`.

before:

```rb
blob.service_url(filename: ActiveStorage::Filename.new("new.txt"))

blob.service_url(filename: "new.txt")
=> NoMethodError: undefined method `parameters' for "new.txt":String

params = {}
blob.service_url(filename: params[:filename])
=> NoMethodError: undefined method `parameters' for nil:NilClass
```

https://github.com/rails/rails/blob/8c5a7fbefd3cad403e7594d0b6a5488d80d4c98e/activestorage/lib/active_storage/service.rb#L124

after:

```rb
blob.service_url(filename: "new.txt")
```

georgeclaghorn commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -110,6 +110,9 @@ def text?
   def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: self.filename, **options)
+    filename ||= self.filename
+    filename = ActiveStorage::Filename.new(filename) if filename.is_a?(String)
+
```
Let’s extract `ActiveStorage::Filename.wrap` and call it below:

```ruby
class ActiveStorage::Filename
  def self.wrap(filename)
    filename.kind_of?(self) ? filename : new(filename)
  end
end
```

huacnlee replied:
done

huacnlee commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -110,6 +110,8 @@ def text?
   # Hiding the +service_url+ behind a redirect also gives you the power to change services without updating all URLs. And
   # it allows permanent URLs that redirect to the +service_url+ to be cached in the view.
   def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: self.filename, **options)
+    filename = ActiveStorage::Filename.wrap(filename || self.filename)
```
`filename || self.filename` for nil case to use default value from `blob.filename`

georgeclaghorn commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -110,6 +110,8 @@ def text?
   # Hiding the +service_url+ behind a redirect also gives you the power to change services without updating all URLs. And
   # it allows permanent URLs that redirect to the +service_url+ to be cached in the view.
   def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: self.filename, **options)
```
Now that we fall back to `self.filename` in the method body, make `nil` the default in the formal argument list to eliminate the duplication:

```ruby
def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: nil, **options)
```

huacnlee replied:
👍

huacnlee committed:
Fix `blob.service_url` for supports string or nil `:filename` option.

- Make sure `blob.service_url` present a `ActiveStorage::Filename` type to `serivce.url`.
- Add `ActiveStorage::Filename.wrap` method.

before:

```rb
blob.service_url(filename: ActiveStorage::Filename.new("new.txt"))

blob.service_url(filename: "new.txt")
=> NoMethodError: undefined method `parameters' for "new.txt":String

params = {}
blob.service_url(filename: params[:filename])
=> NoMethodError: undefined method `parameters' for nil:NilClass
```

after:

```rb
blob.service_url(filename: "new.txt")
blob.service_url(filename: nil)
```

huacnlee commented:
done
-- 
李华顺 (Jason Lee) <http://huacnlee.com>
----------------------------------------
http://huacnlee.com
http://twitter.com/huacnlee
http://github.com/huacnlee
